File: apps/api/src/domain/services/auth-ports.ts
// Capa DOMAIN: contratos de auth. Implementaciones en infrastructure/ — el dominio
// no conoce argon2 ni JWT.

export interface PasswordHasher {
  hash(plain: string): Promise<string>;
  verify(hash: string, plain: string): Promise<boolean>;
}

export interface TokenPair {
  accessToken: string;
  refreshToken: string;
  refreshTokenHash: string;
  refreshExpiresAt: Date;
}

export interface TokenService {
  /** Firma un access token corto (JWT). */
  signAccessToken(userId: string): Promise<string>;
  /** Genera un refresh token opaco aleatorio (devuelve también su hash para almacenar). */
  generateRefreshToken(userId: string): Promise<TokenPair>;
  /** Verifica un access token; devuelve userId o null. */
  verifyAccessToken(token: string): Promise<string | null>;
  /** Hash del refresh token para comparación segura en DB. */
  hashRefreshToken(token: string): string;
}

File: apps/api/internal/domain/services/auth_ports.go
package services

import "time"

type PasswordHasher interface {
	Hash(plain string) (string, error)
	Verify(hash, plain string) (bool, error)
}

type TokenPair struct {
	AccessToken      string    `json:"accessToken"`
	RefreshToken     string    `json:"refreshToken"`
	RefreshTokenHash string    `json:"refreshTokenHash"`
	RefreshExpiresAt time.Time `json:"refreshExpiresAt"`
}

type TokenService interface {
	SignAccessToken(userID string) (string, error)
	VerifyAccessToken(token string) (string, error)
	GenerateRefreshToken(userID string) (TokenPair, error)
	HashRefreshToken(token string) string
}
